src/etwlogger_driver.c

```diff
@@ -113,26 +113,98 @@ static char* lastErrorToString(DWORD lastError)
 }
 
 
-static int isETWLoggerInit = 0;
+static volatile LONG isETWLoggerInit = 0;
 
 static void lazyRegisterEventProvider(void)
 {
     /*lazily init the logger*/
-    if (isETWLoggerInit == 0)
+    if (InterlockedCompareExchange(&isETWLoggerInit, 1, 0)==0)
     {
         ULONG event_register_result = EventRegisterMicrosoft_ServiceBus();
         if (event_register_result != ERROR_SUCCESS)
         {
             /*go back to printf, maybe somebody notices*/
             (void)printf("failure in EventRegisterMicrosoft_ServiceBus: %lu\r\n", event_register_result);
+            (void)InterlockedExchange(&isETWLoggerInit, 0); /*allow a reattempt to register*/
         }
         else
         {
             /*should check if the provider is registered with the system... maybe at a later time*/
-            isETWLoggerInit = 1; /*and stays 1 until the process exits*/ /*sorry, no graceful exit with EventUnregisterMicrosoft_ServiceBus*/
+            /*isETWLoggerInit stays 1 until the process exits*/ /*sorry, no graceful exit with EventUnregisterMicrosoft_ServiceBus*/
             LogInfo("EventRegisterMicrosoft_ServiceBus success"); /*self logging that the log service has started*/
         }
     }
+    else
+    {
+        /*do nothing, already registered/attempted*/
+    }
+}
+
+/*the below interlocked variable initially set to 0*/
+/* At the first error returned by EventWriteLogInfoEvent or by EventWriteLogErrorEvent a printf will be performed and the variable will be switched to 1. */
+/* while the variable is set to "1" no fallback printf will be performed */
+/* at the first non-erroneous EventWriteLogInfoEvent or EventWriteLogErrorEvent the printf fallback will be re-enabled */
+static volatile LONG stopPrintfing = 0;
+
+void perform_EventWriteLogErrorEvent(const char* content, const char* file, const SYSTEMTIME* t, const char* func, int line)
+{
+    ULONG event_write_result = EventWriteLogErrorEvent(content, file, t, func, line);
+    if (event_write_result != ERROR_SUCCESS)
+    {
+        if (InterlockedCompareExchange(&stopPrintfing, 1, 0) == 0)
+        {
+            (void)printf("failed in EventWriteLogErrorEvent: %lu. Further failing calls to EventWriteLog... will not result in printf\r\n", event_write_result);
+        }
+    }
+    else
+    {
+        InterlockedExchange(&stopPrintfing, 0);
```

Comment:
>InterlockedExchange [](start = 8, length = 19)

(void)

Reply:
added to all InterlockedExchanges

---
In reply to: [319620175](https://github.com/Azure/azure-c-shared-utility/pull/367#discussion_r319620175) [](ancestors = 319620175)